lib/user_permissions_controller_methods.rb

```diff
@@ -24,4 +24,16 @@ def zip_permissions(applications, user)
       [application, user.application_permissions.where(application_id: application.id)]
     end
   end
+
+  def all_applications_and_permissions_for(user)
+    permission_names = Hash[SupportedPermission.pluck(:id, :name)]
+    app_permissions = {}
+
+    UserApplicationPermission.where(user: user).includes(:application).find_each do |permission|
+      app_permissions[permission.application] ||= []
+      app_permissions[permission.application] << permission_names[permission.supported_permission_id]
+    end
```

Comment:
If we did `user.supported_permissions.includes(:application).group_by(&:application)` here then I think we could avoid the `permission_names` hash, and let `group_by` do the aggregation for us.  At least I think so, anyway.  The reason I suggest this is that it took me a moment to work out why we had `permission_names` and what the ultimate state of `app_permissions` would be (a hash of application instances as keys, with arrays of permission names as values), and I think leveraging AR methods might help with comprehension.